### PR TITLE
ERC20Gauges

### PR DESCRIPTION
**Pull request by Joeysantoro**
A liquid gauge mechanism embedded inside an ERC-20 for Tribe Launch and a Tribe tokenomic upgrade.

https://hackmd.io/uDcwqBqGRZ-dHIUmzMX8_w for details

### CODE REVIEW

File: src/token/ERC20Gauges.sol [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+        emit IncrementGaugeWeight(msg.sender, gauge, weight);
+    } 
+
+    /** 
+     @notice increment a list of gauges with some weights for the caller
+     @param gaugeList the gauges to increment
+     @param weights the weights to increment by
+     @return newUserWeight the new user weight
+    */
+    function incrementGauges(address[] calldata gaugeList, uint256[] calldata weights) external returns(uint256 newUserWeight) {
```
Comment (ElliotFriedman) [inline src/token/ERC20Gauges.sol:168]: Feels like there is a lot of duplicate logic between incrementGuage and incrementGuages. Would be nice to put some of the logic in a helper function and re-use.

File: src/token/ERC20Gauges.sol [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+        uint256 size = gaugeList.length;
+
+        uint256 totalFree;
+        for (uint256 i = 0; i < size; i++) {
+            address gauge = gaugeList[i];
+            if (!_deprecatedGauges.contains(gauge)) revert InvalidGaugeError();
+
+            uint256 weight = getUserGaugeWeight[user][gauge];
+            if (weight != 0) {
+                getUserGaugeWeight[user][gauge] = 0;
```
Comment (ElliotFriedman) [inline src/token/ERC20Gauges.sol]: delete instead of assign to 0.

File: src/token/ERC20Gauges.sol [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+        address[] memory gaugeList = _gauges.values();
+
+        // Free gauges until through entire list or under weight
+        uint256 size = gaugeList.length;
+        for (uint256 i = 0; i < size && (userFreeWeight + totalFreed) < weight; i++) {
+            address gauge = gaugeList[i];
+            uint256 userGaugeWeight = getUserGaugeWeight[user][gauge];
+            if (userGaugeWeight != 0) {
+                totalFreed += userGaugeWeight;
+                getUserGaugeWeight[user][gauge] = 0;
```
Comment (ElliotFriedman) [inline src/token/ERC20Gauges.sol]: delete instead of assigning to 0

File: src/test/token/ERC20Gauges.t.sol [hunk 4]
```diff
@@ -0,0 +1,512 @@
...
+    /*///////////////////////////////////////////////////////////////
+                        TEST ADMIN GAUGE OPERATIONS
+    //////////////////////////////////////////////////////////////*/
+
+    function testSetMaxGauges() public {
+        token.setMaxGauges(2);
+        require(token.maxGauges() == 2);
+    }
+
+    function testFailSetMaxUnder() public {
```
Comment (thomas-waite) [inline src/test/token/ERC20Gauges.t.sol]: `testFail` is considered an anti-pattern as it is not a strict test as to why a call failed. 

The preferred way to test would be to change the test prefix from `testFail` to `test` and then place before the expected to fail call:

```
vm.expectRevert(bytes("ERROR_REASON_STRING"));
```
where `ERROR_REASON_STRING` is the error that you are expecting to be reported, e.g. `'MaxGaugeError'`.

File: src/test/token/ERC20Gauges.t.sol [hunk 5]
```diff
@@ -0,0 +1,512 @@
...
+        token.setMaxGauges(2);
+        token.addGauge(gauge1);
+        token.setMaxGauges(0);
+    }
+
+    function testFailSetMaxGaugesNonOwner() public {
```
Comment (thomas-waite) [inline src/test/token/ERC20Gauges.t.sol]: Can use `expectRevert()` here too to check the test fails for the expected reason

File: src/test/token/ERC20Gauges.t.sol [hunk 6]
```diff
@@ -0,0 +1,512 @@
...
+        uint[] memory weights = new uint[](3);
+        gaugeList[0] = gauge2;
+        gaugeList[1] = gauge1;
+        weights[0] = 1e18;
+        weights[1] = 2e18;
+
+        token.incrementGauges(gaugeList, weights);    
+    }
+
+    function testDecrement() public {
```
Comment (thomas-waite) [inline src/test/token/ERC20Gauges.t.sol:292]: For some of these more involved tests, I'd perhaps add a /// @notice natspec comment to explain what they're testing specifically. No strong feelings though

File: src/token/ERC20Gauges.sol [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+        getUserWeight[user] -= totalFree;
+    }
+
+    /*///////////////////////////////////////////////////////////////
+                        ADMIN GAUGE OPERATIONS
+    //////////////////////////////////////////////////////////////*/
+
+    /// @notice add a new gauge. Requires auth by `authority`.
+    function addGauge(address gauge) external requiresAuth {
+        if (_gauges.length() >= maxGauges) revert MaxGaugeError();
```
Comment (thomas-waite) [inline src/token/ERC20Gauges.sol:261]: Should we check that a zero address hasn't been accidentally supplied in `addGauge()`? Is that still best practice?

**Review by thomas-waite — COMMENTED** [review]
Looks good to me. 

Really like how `ERC20Gauges.sol` is laid out, it's easy and clear to follow what's happening. Makes a lot of sense.

I think the testing can be made stricter and we can guarantee that the desired failing code paths are being reached by using `vm.expectRevert(bytes('MaxGaugeError'))` etc

File: src/token/ERC20Gauges.sol (hunk 1 above)
Comment (Joeysantoro) [reply src/token/ERC20Gauges.sol:168]: good call, made two helpers each for increment/decrement

File: src/token/ERC20Gauges.sol (hunk 2 above)
Comment (Joeysantoro) [reply src/token/ERC20Gauges.sol]: syntactically those are the same. Delete is most helpful for arrays/structs not basic types

File: src/token/ERC20Gauges.sol (hunk 7 above)
Comment (Joeysantoro) [reply src/token/ERC20Gauges.sol:261]: can go either way but I'll add. Good call!

File: src/test/token/ERC20Gauges.t.sol (hunk 4 above)
Comment (Joeysantoro) [reply src/test/token/ERC20Gauges.t.sol]: great idea, done

File: src/test/token/ERC20Gauges.t.sol (hunk 6 above)
Comment (Joeysantoro) [reply src/test/token/ERC20Gauges.t.sol:292]: added some comments!

File: src/test/token/ERC20Gauges.t.sol [hunk 8]
```diff
@@ -0,0 +1,533 @@
...
+
+    function testSetMaxGauges() public {
+        token.setMaxGauges(2);
+        require(token.maxGauges() == 2);
+    }
+
+    function testSetMaxUnder() public {
+        token.setMaxGauges(2);
+        token.addGauge(gauge1);
+        hevm.expectRevert(abi.encodeWithSignature("MaxGaugeError()"));
```
Comment (thomas-waite) [inline src/test/token/ERC20Gauges.t.sol:30]: You can use `bytes` instead of `abi.encodeWithSignature` I think - I use it

**Review by thomas-waite — APPROVED** [review]
Looks good. One minor convenience thing, you can use `bytes` instead of `abi.encodeWithSignature` in the `expectRevert()` messages afaik

File: src/test/token/ERC20Gauges.t.sol (hunk 8 above)
Comment (Joeysantoro) [reply src/test/token/ERC20Gauges.t.sol:30]: > The error data would be encoded identically as the ABI encoding for function calls, i.e., abi.encodeWithSignature("InsufficientBalance(uint256,uint256)", balance[msg.sender], amount).

https://blog.soliditylang.org/2021/04/21/custom-errors/

**Pull request merged**